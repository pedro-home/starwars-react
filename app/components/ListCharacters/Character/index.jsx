import React from 'react';
import { Panel, List } from 'rsuite';
import Info from './Info';

const Component = ({ data, imageUrl }) => {
  return (
    <Panel shaded bordered bodyFill>
      <img src={imageUrl} height="240" />
      <Panel>
        <List>
          {Object.entries(data).map((key, value) => (
            <List.Item key={key}>
              <Info type={key} value={value} />
            </List.Item>
          ))}
        </List>
      </Panel>
    </Panel>
  );
};

export default Component;
