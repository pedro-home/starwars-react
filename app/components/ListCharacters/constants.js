export const OFFSET_SCROLL = 200;
