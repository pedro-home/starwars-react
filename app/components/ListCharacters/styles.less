.list-characters {
  margin: -1.25rem;

  > * {
    padding: 1.25rem;
  }
}

@import './Character/styles';
