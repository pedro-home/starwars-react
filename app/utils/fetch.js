function fetch(url, options) {
  return window.fetch(url, options).then(response => {
    if (response.ok) {
      let body;
      const contentType = response.headers.get('content-type');
      if (contentType && contentType.indexOf('application/json') !== -1) {
        body = response.json();
      } else {
        body = response.text();
      }

      return body;
    }

    throw new Error(`${response.status}: ${response.statusText}`);
  });
}

export default fetch;
