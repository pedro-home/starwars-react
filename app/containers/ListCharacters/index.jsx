import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { getErrorMessage, getData, isWaiting } from 'providers/StarWars/selectors';
import ListCharacters from 'components/ListCharacters';

const Container = ({ error, data, isLoading }) => {
  return <ListCharacters error={error} data={data} isLoading={isLoading} />;
};

const mapStateToProps = createStructuredSelector({
  error: getErrorMessage(),
  isLoading: isWaiting(),
  data: getData(),
});

export default connect(mapStateToProps)(Container);
