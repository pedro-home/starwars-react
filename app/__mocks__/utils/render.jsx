import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import configureStore from 'store';
import Refs from 'contexts/Refs';
import { DEFAULT_REFS } from 'contexts/Refs/constants';

export function renderWithRedux(component, initialState = {}) {
  const store = configureStore(initialState);
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
}

export function renderWithApp(component, initialState = {}, refsValue = DEFAULT_REFS) {
  return renderWithRedux(<Refs.Provider value={refsValue}>{component}</Refs.Provider>, initialState);
}
