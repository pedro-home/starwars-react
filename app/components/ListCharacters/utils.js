import femaleLogo from './assets/female.jpg';
import maleLogo from './assets/male.jpg';
import hermaphroditeLogo from './assets/hermaphrodite.jpg';
import droidLogo from './assets/droid.jpg';
import { OFFSET_SCROLL } from './constants';

export const getCharacterImage = name => {
  let logo;
  switch (name) {
    case 'female':
      logo = femaleLogo;
      break;
    case 'male':
      logo = maleLogo;
      break;
    case 'hermaphrodite':
      logo = hermaphroditeLogo;
      break;
    default:
      logo = droidLogo;
  }

  return logo;
};

export const hasEndedScroll = (offset = OFFSET_SCROLL) => {
  return window.scrollY + window.innerHeight >= document.body.scrollHeight - offset;
};
