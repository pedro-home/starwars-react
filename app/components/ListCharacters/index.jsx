import React, { useEffect } from 'react';
import { FlexboxGrid, Loader, Col } from 'rsuite';
import PropTypes from 'prop-types';
import Quote from 'components/Quote';
import ContainerCenter from 'components/ContainerCenter';
import Character from './Character';
import { getCharacterImage, hasEndedScroll } from './utils';

const Component = ({ error, data, isLoading, nextFetch, fetch }) => {
  useEffect(() => {
    if (!nextFetch) {
      return;
    }

    const handleScroll = () => {
      if (hasEndedScroll() && !isLoading) {
        fetch(nextFetch);
      }
    };

    window.addEventListener('scroll', handleScroll);

    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, [nextFetch, fetch, isLoading]);

  let component = (
    <>
      <FlexboxGrid className="list-characters">
        {data.map(value => (
          <FlexboxGrid.Item componentClass={Col} md={6} sm={12} colspan={24} key={value.name}>
            <Character data={value} imageUrl={getCharacterImage(value.gender)} />
          </FlexboxGrid.Item>
        ))}
        {isLoading && (
          <FlexboxGrid.Item key="loader" componentClass={Col} colspan={24}>
            <Loader center content="loading..." />
          </FlexboxGrid.Item>
        )}
      </FlexboxGrid>
    </>
  );

  if (error !== '' || (data.length === 0 && !isLoading)) {
    component = (
      <ContainerCenter>
        <Quote text="Search your feelings, you know you to be true" />
      </ContainerCenter>
    );
  }

  return component;
};

Component.defaultProps = {
  data: [],
  isLoading: false,
  fetch: () => {},
  error: '',
};

Component.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object),
  isLoading: PropTypes.bool,
  error: PropTypes.string,
  nextFetch: PropTypes.object,
  fetch: PropTypes.func,
};

export default Component;
