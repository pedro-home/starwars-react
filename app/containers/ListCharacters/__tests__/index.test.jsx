import React from 'react';
import { renderWithApp } from '__mocks__/utils/render';

import ListCharacters from '../index';

describe('Given <ListCharacters />', () => {
  let component;

  describe('When initializes', () => {
    beforeEach(() => {
      const { container } = renderWithApp(<ListCharacters />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
