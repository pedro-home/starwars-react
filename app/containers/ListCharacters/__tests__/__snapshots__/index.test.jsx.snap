// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <ListCharacters /> When initializes Then should match the snapshot 1`] = `
<div>
  <div
    class="container-center"
  >
    <h4
      class="quote"
    >
      <q>
        Search your feelings, you know you to be true
      </q>
    </h4>
  </div>
</div>
`;
